vignettes/higher_order.Rmd

```diff
@@ -0,0 +1,312 @@
+records occurring on or after study day 1.
+
+Here is how you could achieve this using `restrict_derivation()`,
+where the function arguments are passed using `params()` and the restriction
+criteria is given using `filter`:
+
+```{r}
+ae <- ae %>%
+  mutate(TEMP_AESEVN = as.integer(factor(AESEV, levels = c("SEVERE", "MODERATE", "MILD")))) %>%
+  restrict_derivation(
+    derivation = derive_var_extreme_flag,
+    args = params(
+      new_var = AHSEVFL,
+      by_vars = vars(USUBJID),
+      order = vars(TEMP_AESEVN, AESTDY, AESEQ),
+      mode = "first"
+    ),
+    filter = AESTDY >= 1
+  )
+```
+
+```{r, eval=TRUE, echo=FALSE}
+ae %>%
+  arrange(USUBJID, AESTDY, AESEQ, desc(TEMP_AESEVN)) %>%
+  dataset_vignette(
+    display_vars = vars(USUBJID, AEDECOD, AESTDY, AESEQ, AESEV, AHSEVFL)
+  )
+```
+
+# Slice Derivation
+
+This function in a way combines the features of the above two. It allows a
+derivation to be applied with different arguments for different slices (subsets)
+of records from the input dataset. You could do this with separate `restrict_derivation()`
+calls for each different set of records, but `slice_derivation()` allows
+to achieve this in one call.
+
+An example would be if you wanted to achieve the same derivation as above
+for records occurring on or after study day 1, but for pre-treatment AEs
+you wanted to flag only the last occurring AE.
+
+Here is how you could achieve this using `slice_derivation()`,
+where the function arguments are passed using `params()` and via the different
+slices controlled by `filter`:
+
+```{r}
+ae <- ae %>%
+  slice_derivation(
+    derivation = derive_var_extreme_flag,
+    args = params(
+      new_var = AHSEV2FL,
+      by_vars = vars(USUBJID)
+    ),
+    derivation_slice(
+      filter = AESTDY >= 1,
+      args = params(order = vars(TEMP_AESEVN, AESTDY, AESEQ), mode = "first")
+    ),
+    derivation_slice(
+      filter = TRUE,
+      args = params(order = vars(AESTDY, AESEQ), mode = "last")
+    )
+  )
+```
+
+```{r, eval=TRUE, echo=FALSE}
+ae %>%
+  arrange(USUBJID, AESTDY, AESEQ, desc(TEMP_AESEVN)) %>%
+  dataset_vignette(
+    display_vars = vars(USUBJID, AEDECOD, AESTDY, AESEQ, AESEV, AHSEV2FL)
+  )
+```
+
+As you can see in the example, the `derivation_slice` ordering is important.
+Here we addressed all the AEs on or after study day 1 first, and then we used
+`filter = TRUE` option to catch all remaining records (in this case
+pre-treatment AEs).
+
+The ordering is perhaps shown even more when we look at the below example where
+three slices are taken. Remember that observations that match with more than one
+slice are only considered for the first matching slice. So in this case we're
+creating a flag for each patient for the record with the first severe AE, and
+then the first moderate AE, and finally flagging the last occurring AE where
+not severe or moderate.
+
+```{r}
+ae <- ae %>%
+  slice_derivation(
+    derivation = derive_var_extreme_flag,
```

Comment:
Ok I think the example is clear, only one function with different arguments. But still might be good to mention

Reply:
sure, i added the word "single" in select places to make this clearer - let me know if you prefer any more wording to be added